### DISCUSSION TITLE
DataTable: keep filtered list an origin list synchronized on sorting

### DISCUSSION THREAD

**Discussion by hgfrey** [primefaces/community#873] [PrimeFaces]
If I first filter and then sort in a dataTable, the original list doesn't notice anything from the sorting. But I always have to validate using the original list. If I now remove the filtering and carry out the validation, the order of the error messages does not match the order in the table.

Is there any way to keep synchronized filtered list and origin list on sorting?

**Comment by hgfrey** [comment]
Das ist es, was ich jetz tmache um die Sortierung für die gesamte Liste einer Tabelle zu übernehmen, wenn ein Filter verwendet wurde:

>         SortingInformation sortingInformation = this.sessionInfo.getSortingInformation();
>         if (sortingInformation != null && this.listFiltered != null && !this.listFiltered.isEmpty()) {
>             String sortBy = sortingInformation.getSortBy();
>             int sortOrder = sortingInformation.getSortOrder();
>             int nullSortOrder = sortingInformation.getNullSortOrder();
>             ComparingValues comparingValues = new ComparingValues();
>             listActive.sort((l1, l2) -> {
>                 try {
>                     Object val1 = PropertyUtils.getProperty(l1, sortBy);
>                     Object val2 = PropertyUtils.getProperty(l2, sortBy);
>                     int i = comparingValues.compare(val1, val2, nullSortOrder);
>                     return i * sortOrder;
>                 } catch (Exception e) {
>                     throw new RuntimeException(e);
>                 }
>             });
>         }
That looks relatively compact. What cannot be seen here are the comparison routines, which have to take into account the different data types, as well as ascending and descending and also the positioning of zero values.

What's really unpleasant is that I need two events to synchronize the list. The first gives me the sorting criteria and the second signals me the end of sorting in the UI so I can start sorting my list.

>     <p:ajax event="sort" listener="#{ptPlanungController.onSort}" global="false"/>
>     <f:event type="org.primefaces.event.data.PostSortEvent" listener="#{ptPlanungController.onSortDatenzeilen}"/>
> 
>     public void onSort(SortEvent event) {
>         sessionInfo.setSortingInformation(null);
>         for (Map.Entry<String, SortMeta> entry : event.getSortBy().entrySet()) {
>             SortMeta sortMeta = entry.getValue();
>             if (sortMeta.getPriority() == 0) {
>                 sessionInfo.setSortingInformation(new SortingInformation(sortMeta.getField(), sortMeta.getOrder().intValue(), sortMeta.getNullSortOrder()));
>                 break;
>             }
>         }
>     }
> 
>     public void onSortDatenzeilen(ComponentSystemEvent event) {
>         synchronizeListActive();
>     }
 
I have to remember the sorting information from the first call in a SessionInfo object so that I can use it in the second call. Why can't I also get the sorting information on the second call? That would be a good idea at a sort event, right?

**Reply by melloware** [reply]
i would need a Test reproducer. I am unclear what you are trying to do...

Please provide an executable example using the [PrimeFaces Test](https://github.com/primefaces/primefaces-test) project.  It is the only way developers can debug your problem to help.

**Reply by melloware** [reply]
Will close for now.
